cluster_generator/data_structures.py

```diff
@@ -1,100 +1,310 @@
+    box_size,
+    left_edge,
+    domain_dimensions,
+    chunking=True,
+    chunksize=64,
+    center=(0, 0, 0),
+):
+    """
+    Construct the radial coordinate array for a given grid.
+
+    Parameters
+    ----------
+    io_array: :py:class:`np.ndarray`
+        The array into which the output data should be written.
+    box_size: tuple or list
+        A length 3 iterable with the values of each of the box sizes.
+    left_edge: tuple or list
+        A lenth 3 iterable containing the coordinates of the bottom left most point on the grid.
+    domain_dimensions: tuple or list
+        The number of "cells" to place on each axis of the grid.
+    chunking: bool, optional
+        (Default, ``True``) If enabled, chunking is used to reduce the memory load of the computation.
+    chunksize: int, optional
+        (Default, ``64``) The maximum size of a given chunk.
+    center: tuple, optional
+        (Default, ``(0,0,0)``) The center point of the chosen coordinate system.
+
+    Returns
+    -------
+    :py:class:`np.ndarray`
+        The boundary box for the output geometry. The geometry array itself is written to the ``io_array`` and not returned.
+    """
+    mylog.info(
+        f"Constructing grid geometry on domain {domain_dimensions}, chunking = {chunking}"
+    )
+    domain_dimensions = np.array(domain_dimensions)
+    if chunking:
+        exp_mem = 1240 * np.prod(domain_dimensions / chunksize) / (8 * 1e6)
+        mylog.info(
+            f"Expected memory usage / iteration is {np.round(exp_mem,decimals=2)} MB."
+        )
+    else:
+        exp_mem = 256 * np.prod(domain_dimensions) / (8 * 1e6)
+        mylog.info(f"Expected memory usage is {np.round(exp_mem,decimals=2)} MB.")
+
+    bbox = np.array(
+        [
+            [left_edge[0], left_edge[0] + box_size],
+            [left_edge[1], left_edge[1] + box_size],
+            [left_edge[2], left_edge[2] + box_size],
+        ]
+    )
+
+    if chunking:
+        # compute iteratively.
+        _chunk_ids, _chunk_coords = _build_chunks(chunksize, domain_dimensions)
+
+        for chunk in tqdm(
+            _chunk_ids,
+            desc="Constructing coordinate grids...",
+            leave=False,
+            disable=(not cgparams["system"]["display"]["progress_bars"]),
+        ):
+            _cc = _chunk_coords[:, :, *chunk]
+            x, y, z = np.mgrid[
+                _cc[0, 0] : _cc[1, 0], _cc[0, 1] : _cc[1, 1], _cc[0, 2] : _cc[1, 2]
+            ] * ((bbox[:, 1] - bbox[:, 0]) / (domain_dimensions - 1)).reshape(
+                3, 1, 1, 1
+            ) + bbox[
+                :, 0
+            ].reshape(
+                3, 1, 1, 1
+            )
+            rr = np.sqrt(
+                (x - center[0]) ** 2 + (y - center[1]) ** 2 + (z - center[2]) ** 2
+            )
+
+            io_array[
+                _cc[0, 0] : _cc[1, 0], _cc[0, 1] : _cc[1, 1], _cc[0, 2] : _cc[1, 2]
+            ] = rr
+
+        del x, y, z, rr
+        gc.collect()
```

Comment:
@Eliza-Diggins are you sure we need to garbage-collect by hand here? It's very rare that one needs to call this explicitly in packages. It usually indicates that we should re-write things to avoid it.

Reply:
@jzuhone I don't think its strictly necessary.